files/default/neutron-ha-tool.py

```diff
@@ -576,6 +600,199 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+        LOG.error("%d errors encountered during evacuation", errors)
+
+    return errors
+
+
+def migrate_dhcp_networks_from_agent(qclient, agent, targets, agent_picker,
+                                     noop, wait_for_network, fail_fast):
+    LOG.info("Querying agent_id=%s for networks to migrate away", agent['id'])
+
+    networks_on_agent = qclient.list_networks_on_dhcp_agent(agent['id'])['networks']
+
+    agent_picker.set_agents(targets)
+
+    migrations = 0
+    errors = 0
+    for network in networks_on_agent:
+        target = agent_picker.pick()
+
+        if noop:
+            LOG.info("Would try to migrate network=%s from agent=%s "
+                 "to agent=%s", network['id'], agent['id'], target['id'])
+            migrations += 1
+            continue
+
+        try:
+            migrate_dhcp_network(qclient, network, agent, target, wait_for_network)
+            migrations += 1
+        except:
+            LOG.exception("Failed to migrate network_id=%s from agent=%s to "
+                          "agent=%s", network['id'], agent['id'], target['id'])
+            errors += 1
+            if fail_fast:
+                break
+
+    return (migrations, errors)
+
+
+def migrate_dhcp_network(qclient, network, agent, target,
+                         wait_for_network=True):
+    """
+    Returns nothing, and raises exceptions on errors.
+
+    :param qclient: A neutronclient
+    :param network: The network to migrate
+    :param agent: The DHCP agent to migrate from
+    :param target: The DHCP agent to migrate to
+    :param wait_for_network: Optional wait-for-dhcp-network flag
+    """
+
+    LOG.info("Migrating network=%s from agent=%s to agent=%s",
+             network['id'], agent['id'], target['id'])
+
+    # Prevent SIGTERM only when modifying data through the API, and allow it
+    # when ensuring the API operations completed successfully
+    with term_disabled():
+
+        # Remove the network from the original agent
+        qclient.remove_network_from_dhcp_agent(agent['id'], network['id'])
+        LOG.debug("Removed network from agent=%s" % agent['id'])
+
+        # ensure it is removed
+        if dhcp_network_is_on_agent(qclient, network['id'], agent['id']):
+            raise RuntimeError("Failed to remove network_id=%s from "
+                               "agent_id=%s" % (network['id'], agent['id']))
+
+        # TODO: this is copied from router evac, is this the same for dhcp?
+        # Only wait for the ports if the source agent is really alive,
+        # otherwise we won't see any state change on dhcp ports.
+        if wait_for_network and agent['alive']:
+            wait_dhcp_network_migrated(qclient, network['id'], agent['host'],
+                                       state="DOWN")
+
+        # TODO: what to do if target already has the network?
+        # skip? pick another target? remote and re-add?
+        if not dhcp_network_is_on_agent(qclient, network['id'], target['id']):
```

Comment:
Why the check again? In this part of code `dhcp_network_is_on_agent` should already return false.... or do you think something could have happened during `wait_dhcp_network_migrated`?